web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php

```diff
@@ -0,0 +1,269 @@
+        $requested_dependencies,
+        $this->solutionData[$solution_id]['dependencies'],
+      ]);
+    }
+
+    $releases = $requested_dependencies['asset_release'] ?: [];
+    unset($requested_dependencies['asset_release']);
+
+    $return = $solution_ids + $releases;
+    foreach ($requested_dependencies as $ids_per_bundle) {
+      $return += $ids_per_bundle;
+    }
+
+    // Remove skipped solution ids.
+    $return = array_diff($return, $skipped_solutions);
+    return $return;
+  }
+
+  /**
+   * Sets the solution category to the solution data.
+   *
+   * @param string $solution_id
+   *   The solution id.
+   * @param string $category
+   *   The solution category.
+   */
+  protected function setSolutionCategory(string $solution_id, string $category): void {
+    $this->ensureEntityDataLoaded();
+    $this->solutionData[$solution_id]['category'] = $category;
+  }
+
+  /**
+   * Returns a list of solutions that have been marked as unchanged.
+   *
+   * @param string $category
+   *   The category to filter by.
+   *
+   * @return array
+   *   An array of solution ids.
+   */
+  protected function getSolutionIdsMatchingCategory(string $category): array {
+    $this->ensureEntityDataLoaded();
+    $return = [];
+    foreach ($this->solutionData as $solution_id => $solution_data) {
+      if ($solution_data['category'] === $category) {
+        $return[$solution_id] = $solution_id;
+      }
+    }
+    return $return;
+  }
+
+  /**
+   * Retrieves the solution category from the persistent state.
+   *
+   * @param string $solution_id
+   *   The solution id.
+   *
+   * @return string
+   *   The solution category.
+   *
+   * @throws \Exception
+   *   Thrown if the category of the solution requested has not been set yet.
+   */
+  protected function getSolutionCategory(string $solution_id): string {
+    $this->ensureEntityDataLoaded();
+    if (empty($this->solutionData[$solution_id]['category'])) {
+      throw new \Exception("Category has not been set for solution with id {$solution_id}");
+    }
+    return $this->solutionData[$solution_id]['category'];
+  }
+
+  /**
+   * Returns whether a solution is listed in the given category.
+   *
+   * @param string $solution_id
+   *   The solution entity id.
+   * @param string $category
+   *   The category to check.
+   *
+   * @return bool
+   *   Whether the solution is listed under this category.
+   */
+  protected function solutionHasCategory(string $solution_id, string $category): bool {
+    $this->ensureEntityDataLoaded();
+    return $this->solutionData[$solution_id]['category'] === $category;
+  }
+
+  /**
+   * Returns a hash related to a given entity.
+   *
+   * @param string $entity_id
+   *   The entity id.
+   *
+   * @return string
+   *   The hash related to the entity.
+   */
+  protected function getEntityHash(string $entity_id): string {
+    $this->ensureEntityDataLoaded();
+    return $this->entityHashes[$entity_id] ?? '';
```

Comment:
The case would be when the user would request the hash of an entity before it is calculated. Generally, the two cases we have in our code is quite caseproof that it will always contain the entities. 
However, I do agree that it is better to simply return the hash and in case there is none, php will throw a warning/error about undefined index. Removing it.